app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,255 @@
+        self.assertEqual(len(res.data), 0)
+
+    # GET /recipes/<id>/
+    def test_get_recipe_by_id_success(self):
+        """Test retrieving a recipe by ID"""
+        # Given
+        sample_recipe()
+        recipe = sample_recipe({
+            'name': 'Turkey bubble & squeak',
+            'description': 'An easy, lean, recipe'
+        })
+        recipe.ingredients.add(sample_ingredient(recipe=recipe, name='Turkey'))
+
+        serializer = RecipeSerializer(recipe)
+
+        # When
+        url = recipe_url(recipe.id)
+        res = self.client.get(url)
+
+        # Then
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_get_recipe_by_id_not_found(self):
+        """Test retrieving a recipe by ID when the recipe doesn't exist"""
+        # Given
+        # No recipes defined
+
+        # When
+        res = self.client.get(recipe_url(recipe_id=1))
+
+        # Then
+        self.assertEqual(res.status_code, status.HTTP_404_NOT_FOUND)
+
+    # POST /recipes/
+    def test_create_recipe(self):
+        """Test creating a new recipe"""
+        # Given
+        payload = {
+            'name': 'Pizza',
+            'description': 'Put it in the oven',
+            'ingredients': [
+                {'name': 'cheese'},
+                {'name': 'dough'},
+                {'name': 'tomato'},
+            ]
+        }
+
+        # When
+        res = self.client.post(RECIPES_URL, payload)
+
+        # Then
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+
+        recipe = Recipe.objects.get(id=res.data['id'])
+        ingredients = recipe.ingredients.all()
+
+        self.assertEqual(ingredients.count(), 3)
+
+    def test_create_recipe_invalid_request(self):
+        """Test creating a new recipe with invalid payload"""
+        # Given
+        payload = {
+            'name': '',
+        }
+
+        # When
+        res = self.client.post(RECIPES_URL, payload)
+
+        # Then
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    # PATCH /recipes/<id>/
+    def test_patch_recipe(self):
+        """Test partial update of a recipe"""
+        # Given
+        recipe = sample_recipe()
+
+        ingredient1 = sample_ingredient(recipe, name='Sweetcorn')
+        ingredient2 = sample_ingredient(recipe, name='Ham')
+
+        recipe.ingredients.add(ingredient1, ingredient2)
+
+        # When
+        new_ingredient_name = 'Pepperoni'
+        payload = {
+            'name': recipe.name,
+            'description': recipe.description,
+            'ingredients': [
+                {'name': new_ingredient_name}
+            ]
+        }
+
+        res = self.client.patch(recipe_url(recipe_id=recipe.id), payload)
+
+        # Then
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(len(ingredients), 1)
```

Comment:
Instead of `.all()` on the query set + `len()` here you can replace `all()` with `count()` - it will make a single query returning the count value instead of fetching the objects from the DB - much more efficient / clear. [You can review other query set methods here](https://docs.djangoproject.com/en/3.2/ref/models/querysets/)

Reply:
Thank you, I'll replace Line 225 to have `ingredients.count()` but I'll keep the `.all()` on line 224 as it's needed for an assertion on line 226.